exonum/src/keys/mod.rs

```diff
@@ -0,0 +1,270 @@
+fn validate_file_mode(mode: u32) -> Result<(), Error> {
+    // Check that group and other bits are not set.
+    if (mode & 0o_077) == 0 {
+        Ok(())
+    } else {
+        Err(Error::new(ErrorKind::Other, "Wrong file's mode"))
+    }
+}
+
+/// Struct containing all validator key pairs.
+#[derive(Debug, Clone, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Keys {
+    consensus: KeyPair,
+    service: KeyPair,
+}
+
+impl Keys {
+    /// Create validator keys from provided keypairs.
+    ///
+    /// # Panics
+    ///
+    /// If public key in any keypair doesn't match with corresponding private key.
+    pub fn from_keys(
+        consensus_pk: PublicKey,
+        consensus_sk: SecretKey,
+        service_pk: PublicKey,
+        service_sk: SecretKey,
+    ) -> Self {
+        Self {
+            consensus: (consensus_pk, consensus_sk).into(),
+            service: (service_pk, service_sk).into(),
+        }
+    }
+}
+
+impl Keys {
+    /// Consensus public key.
+    pub fn consensus_pk(&self) -> PublicKey {
+        self.consensus.public_key()
+    }
+
+    /// Consensus private key.
+    pub fn consensus_sk(&self) -> &SecretKey {
+        &self.consensus.secret_key()
+    }
+
+    /// Service public key.
+    pub fn service_pk(&self) -> PublicKey {
+        self.service.public_key()
+    }
+
+    /// Service secret key.
+    pub fn service_sk(&self) -> &SecretKey {
+        &self.service.secret_key()
+    }
+}
+
+fn save_master_key<P: AsRef<Path>, W: AsRef<[u8]>>(
+    path: P,
+    pass_phrase: W,
+    key: &secret_tree::Seed,
+) -> Result<(), Error> {
+    let encrypted_key = EncryptedMasterKey::encrypt(key, pass_phrase)?;
+    let file_content =
+        toml::to_string_pretty(&encrypted_key).map_err(|e| Error::new(ErrorKind::Other, e))?;
+    let mut open_options = OpenOptions::new();
+    open_options.create(true).write(true);
+    // By agreement we use the same permissions as for SSH private keys.
+    #[cfg(unix)]
+    open_options.mode(0o_600);
+    let mut file = open_options.open(path.as_ref())?;
+    file.write_all(file_content.as_bytes())?;
+
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize)]
+struct EncryptedMasterKey {
+    key: ErasedPwBox,
+}
+
+impl EncryptedMasterKey {
+    fn encrypt(
+        key: &secret_tree::Seed,
+        pass_phrase: impl AsRef<[u8]>,
+    ) -> Result<EncryptedMasterKey, Error> {
+        let mut rng = thread_rng();
+        let mut eraser = Eraser::new();
+        eraser.add_suite::<Sodium>();
+        let pwbox = Sodium::build_box(&mut rng)
+            .seal(pass_phrase, key)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't create a pw box"))?;
+        let encrypted_key = eraser
+            .erase(&pwbox)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't convert a pw box"))?;
+
+        Ok(EncryptedMasterKey { key: encrypted_key })
+    }
+
+    fn decrypt(self, pass_phrase: impl AsRef<[u8]>) -> Result<secret_tree::Seed, Error> {
```

Comment:
`SensitiveData` field and methods are private, i can't create instance of it. Is it possible to make them public, or it will break something?

Reply:
You don't need access to `SensitiveData` internal structure, though, since it implements `Deref`:

```patch
@@ -14,7 +14,7 @@
 
 use crate::crypto::{gen_keypair_from_seed, PublicKey, SecretKey, Seed, SEED_LENGTH};
 use failure::format_err;
-use pwbox::{sodium::Sodium, ErasedPwBox, Eraser, Suite};
+use pwbox::{sodium::Sodium, ErasedPwBox, Eraser, SensitiveData, Suite};
 use rand::thread_rng;
 use secret_tree::{Name, SecretTree};
 
@@ -128,20 +128,16 @@
         Ok(EncryptedMasterKey { key: encrypted_key })
     }
 
-    fn decrypt(self, pass_phrase: impl AsRef<[u8]>) -> Result<secret_tree::Seed, Error> {
+    fn decrypt(self, pass_phrase: impl AsRef<[u8]>) -> Result<SensitiveData, Error> {
         let mut eraser = Eraser::new();
         eraser.add_suite::<Sodium>();
         let restored = eraser
             .restore(&self.key)
             .map_err(|_| Error::new(ErrorKind::Other, "Couldn't restore a secret key"))?;
         assert_eq!(restored.len(), SEED_LENGTH);
-        let seed_bytes = restored
+        restored
             .open(pass_phrase)
-            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't open an encrypted key"))?;
-
-        let mut seed: [u8; 32] = [0; 32];
-        seed.copy_from_slice(&seed_bytes[..]);
-        Ok(seed)
+            .map_err(|_| Error::new(ErrorKind::Other, "Couldn't open an encrypted key"))
     }
 }
 
@@ -222,7 +218,7 @@
         let decrypted_seed = key
             .decrypt(pass_phrase)
             .expect("Couldn't decrypt master key ");
-        assert_eq!(seed, &decrypted_seed);
+        assert_eq!(&seed[..], &decrypted_seed[..]);
     }
 
     #[test]
@@ -249,7 +245,7 @@
         let seed = keys.decrypt(pass_phrase).expect("Couldn't decrypt key");
 
         assert_eq!(
-            hex::encode(&seed),
+            hex::encode(&*seed),
             "a05a82575d5f9d1f9469df31896f5b3c14ec4d18b3948cd7c8b09a7eed48b4e0"
         );
     }
```

Reply:
Got it! I missed that RestoredPwBox::open already returns Result<SensitiveData, ..>